[docs] Improved the 'Recipes/React' documentation (#2464)
statelyai/xstate: docs/recipes/react.md

# Usage with React

XState can be used with React to:

> - Co-ordinate local state

Reviewer: ```suggestion
- Coordinate local state
```

American spelling? 😜

Author: 😂  nope, just wrong spelling

Had my grammar corrected by an American - bad form, Matthew

- Manage global state performantly
- Consume data from other hooks

At [Stately](https://stately.ai), we love this combo. It's our go-to stack for creating internal applications.

> To ask for help, check out the `#react-help` channel on [Discord](https://discord.com/invite/xstate).

Reviewer: ```suggestion
To ask for help, check out the [`#react-help` channel in our Discord community](https://discord.gg/vedXj62MfQ).
```

Channel-specific invite link (never expires)

> There are have two versions of 'context': XState's [context](../guides/context.md) and React's context. It's a little confusing!

### Context Provider

React context can be a tricky tool to work with - if you pass in values which change too often, it can result in re-renders all the way down the tree. That means we need to pass in values which change as little as possible.

Luckily, XState gives us a first-class way to do that: `useInterpret`.

```js
import React, { createContext } from 'react';
import { useInterpret } from '@xstate/react';
import { authMachine } from './authMachine';

export const GlobalStateContext = createContext({});

export const GlobalStateProvider = (props) => {
  const authService = useInterpret(authMachine);

  return (
    <GlobalStateContext.Provider value={{ authService }}>
      {props.children}
    </GlobalStateContext.Provider>
  );
};
```

Using `useInterpret` returns a service, which is a static reference to the running machine which can be subscribed to. This value never changes, so we don't need to worry about wasted re-renders.

> ### Utilising context

Reviewer: ```suggestion
### Utilizing context
```

Further down the tree, you can subscribe to the service like this:

```js
import React, { useContext } from 'react';
import { GlobalStateContext } from './globalState';
import { useActor } from '@xstate/react';

export const SomeComponent = (props) => {
  const globalServices = useContext(GlobalStateContext);
  const [state] = useActor(globalServices.authService);

  return state.matches('loggedIn') ? 'Logged In' : 'Logged Out';
};
```

The `useActor` hook listens for whenever the service changes, and updates the state value.

### Improving Performance

There's an issue with the implementation above - this will update the component for any change to the service. Tools like [Redux](https://redux.js.org) use [`selectors`](https://redux.js.org/usage/deriving-data-selectors) for deriving state. Selectors are functions which restrict which parts of the state can result in components re-rendering.

Fortunately, XState exposes the `useSelector` hook.

```js
import React, { useContext } from 'react';
import { GlobalStateContext } from './globalState';
import { useSelector } from '@xstate/react';

> const selector = (state) => {

Reviewer: ```suggestion
const loggedInSelector = (state) => {
```

  return state.matches('loggedIn');
};

export const SomeComponent = (props) => {
  const globalServices = useContext(GlobalStateContext);

>   const isLoggedIn = useSelector(globalServices.authService, selector);

Reviewer: ```suggestion
  const isLoggedIn = useSelector(globalServices.authService, loggedInSelector);
```
